### PR TITLE
[FIX] add nilearn tags to estimators

### LINKED ISSUE #4931: :rotating_light: NIGHTLY DEPENDENCIES TEST: failure on `refs/heads/main`
The run `12311489184` of the workflow testing Nilearn with the nightly build of its dependencies failed on `refs/heads/main`.

You can view the report here: https://github.com/nilearn/nilearn/actions/runs/12311489184

Pinging @nilearn/core

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes #4931 

Changes proposed in this pull request:

- adds a way to add tags to estimators to specify if they accept nifti and / or surface as input
- skip some sklearn estimator checks if that's the case

### CODE REVIEW

File: nilearn/decoding/decoder.py [hunk 1]
```diff
@@ -1135,9 +1135,14 @@ def __sklearn_tags__(self):
         # see https://github.com/scikit-learn/scikit-learn/pull/29677
         ver = parse(sklearn_version)
         if ver.release[1] < 6:
-            return {"require_y": True}
+            from nilearn._utils.class_inspect import tags
+
+            return tags(require_y=True)
+        from nilearn._utils.class_inspect import InputTags
```
Comment (Remi-Gau) [inline nilearn/decoding/decoder.py:1141]:
```suggestion

        from nilearn._utils.class_inspect import InputTags
```

File: nilearn/connectome/connectivity_matrices.py [hunk 2]
```diff
@@ -477,6 +477,13 @@ def _check_input(self, X, confounds=None):
                 f"You provided {confounds.__class__}"
             )
 
+    def _more_tags(self):
+        return super()._more_tags()
+
+    def __sklearn_tags__(self):
+        tags = super().__sklearn_tags__()
+        return tags
+
```
Comment (Remi-Gau) [inline nilearn/connectome/connectivity_matrices.py:486]: double check if this is needed

File: nilearn/connectome/group_sparse_cov.py [hunk 3]
```diff
@@ -597,6 +597,12 @@ def __init__(
         self.memory_level = memory_level
         self.verbose = verbose
 
+    def _more_tags(self):
+        return super()._more_tags()
+
+    def __sklearn_tags__(self):
+        return super().__sklearn_tags__()
+
```
Comment (Remi-Gau) [inline nilearn/connectome/group_sparse_cov.py:605]: double check if this is needed

File: nilearn/connectome/group_sparse_cov.py [hunk 4]
```diff
@@ -1070,6 +1076,12 @@ def __init__(
         self.debug = debug
         self.early_stopping = early_stopping
 
+    def _more_tags(self):
+        return super()._more_tags()
+
+    def __sklearn_tags__(self):
+        return super().__sklearn_tags__()
+
```
Comment (Remi-Gau) [inline nilearn/connectome/group_sparse_cov.py:1084]: double check if this is needed

File: nilearn/_utils/class_inspect.py [hunk 5]
```diff
@@ -126,3 +161,55 @@ class 'cls' and returns the value for these parameters in object
...
 
     return params
+
+
+ver = parse(sklearn_version)
+if ver.release[1] >= 6:
+    from sklearn.utils import InputTags as SkInputTags
+
+    @dataclass
+    class InputTags(SkInputTags):
```
Comment (Remi-Gau) [inline nilearn/_utils/class_inspect.py:171]: will extract this in tags module

File: nilearn/decoding/searchlight.py [hunk 6]
```diff
@@ -328,6 +330,25 @@ def __init__(
...
+        if ver.release[1] < 6:
+            from nilearn._utils.tags import tags
+
+            return tags()
+
+        from nilearn._utils.tags import InputTags
+
+        tags = super().__sklearn_tags__()
+        tags.input_tags = InputTags()
+        return tags
```
Comment (Remi-Gau) [inline nilearn/decoding/searchlight.py:350]: I have had to add this pattern to almost all our estimators.
it's a bit ugly but it gets the job done

**Comment by Remi-Gau** [comment]
Tested locally with sklearn 1.5.2, 1.6.0, 1.7.0dev
Tested in CI with the oldest sklearn we support and also 1.6.0, 1.7.0dev

So unless I missed something obvious things should be all green after merge.

File: nilearn/_utils/class_inspect.py [hunk 7]
```diff
@@ -40,22 +55,25 @@
...
+    So this is a home made generator that yields an estimator instance
     along with a
     - valid check from sklearn: those should stay valid
     - or an invalid check that is known to fail.
 
     If new 'valid' checks are added to scikit-learn,
     then tests marked as xfail will start passing.
 
+    If estimator have some nilearn specific tags
+    then some checks will skipped rather than yielded.
```
Comment (bthirion) [inline nilearn/_utils/class_inspect.py:75]:
```suggestion
    then some checks will skip rather than yield.
```

File: nilearn/_utils/tags.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
+"""Nilearn tags for estimators."""
```
Comment (bthirion) [inline nilearn/_utils/tags.py:1]: Maybe give a short explanation or reference to explain what tags are and why they're needed

File: nilearn/_utils/tests/test_tags.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+        tags = super().__sklearn_tags__()
+        tags.input_tags = InputTags(surf_img=True, niimg_like=False)
+        return tags
+
+
+def test_nilearn_tags():
+    """Check that adding tags to Nilearn estimators work as expected.
+
+    Especially with different skleanr versions.
```
Comment (bthirion) [inline nilearn/_utils/tests/test_tags.py:32]:
```suggestion
    Especially with different sklearn versions.
```

File: nilearn/decoding/searchlight.py [hunk 10]
```diff
@@ -328,6 +330,25 @@ def __init__(
         self.cv = cv
         self.verbose = verbose
 
+    def _more_tags(self):
+        return self.__sklearn_tags__()
+
+    def __sklearn_tags__(self):
+        # TODO
```
Comment (bthirion) [inline nilearn/decoding/searchlight.py:337]: Include a one-line docstring here

**Review by bthirion — COMMENTED** [review]
LGTM

File: nilearn/maskers/surface_labels_masker.py
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/maskers/surface_labels_masker.py:1]: yeah that's because we were not testing the compliance of the surface label masker, but you fixed that in https://github.com/nilearn/nilearn/pull/4921

😉

[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/maskers/surface_labels_masker.py:1]: so after a rebase it should be fine

File: nilearn/decoding/decoder.py [hunk 11]
```diff
@@ -1473,11 +1482,16 @@ def __sklearn_tags__(self):
...
+            return tags(multioutput=True)
+
+        from nilearn._utils.tags import InputTags
+
         tags = super().__sklearn_tags__()
-        if ver.release[1] >= 6:
-            tags.estimator_type = "regressor"
+        tags.estimator_type = "regressor"
         tags.target_tags.required = True
+        tags.input_tags = InputTags()
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/decoding/decoder.py:1494]: good thinking!!!
let me try this then.

[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/decoding/decoder.py]: OK that lead to some wonky infinite recursion so will not it now: maybe we can try to refactor later

**Comment by Remi-Gau** [comment]
@bthirion @man-shu 
the number of lines changed increased quite a bit but this is mostly adding the same doc string to all the `_more_tags` and `__sklearn_tags__`

File: nilearn/decoding/decoder.py [hunk 12]
```diff
@@ -1298,11 +1311,34 @@ def __init__(
...
+
+        from nilearn._utils.tags import InputTags
+
         tags = super().__sklearn_tags__()
-        ver = parse(sklearn_version)
-        if ver.release[1] >= 6:
-            tags.estimator_type = "classifier"
+        tags.estimator_type = "classifier"
+        tags.target_tags.required = True
+        tags.input_tags = InputTags(niimg_like=True, surf_img=True)
```
Comment (man-shu) [inline nilearn/decoding/decoder.py:1341]: Maybe I don't fully understand how these tag-related methods work, but do we still need to redefine them here if they are already in `_BaseDecoder` (which the Decoder object inherits)?

File: nilearn/decoding/decoder.py [hunk 13]
```diff
@@ -1464,20 +1500,33 @@ def __init__(
...
+            return tags(multioutput=True, niimg_like=True, surf_img=True)
+
+        from nilearn._utils.tags import InputTags
+
         tags = super().__sklearn_tags__()
-        if ver.release[1] >= 6:
-            tags.estimator_type = "regressor"
+        tags.estimator_type = "regressor"
         tags.target_tags.required = True
+        tags.input_tags = InputTags(niimg_like=True, surf_img=True)
```
Comment (man-shu) [inline nilearn/decoding/decoder.py:1529]: Same question for the DecoderRegressor?

Comment (man-shu) [reply nilearn/decoding/decoder.py:1529]: Also I am realizing that we haven't updated the docstring for DecoderRegressor about support for Surface mask/images etc. For another PR...

File: nilearn/maskers/base_masker.py [hunk 14]
```diff
@@ -205,6 +206,33 @@ def transform_single_imgs(
...
+        # bumping sklearn_version > 1.5
+        if SKLEARN_LT_1_6:
+            from nilearn._utils.tags import tags
+
+            return tags()
+
+        from nilearn._utils.tags import InputTags
+
+        tags = super().__sklearn_tags__()
+        tags.input_tags = InputTags()
```
Comment (man-shu) [inline nilearn/maskers/base_masker.py:233]: For example, here we only define the tags for the BaseMasker and not for the classes that inherit it (all the NiftiMaskers).

**Comment by Remi-Gau** [comment]
> We'll be happy to drop support for sklearn 1.5...

you read my mind!!!
